aten/src/ATen/core/intrusive_ptr.h

```diff
@@ -0,0 +1,532 @@
+#pragma once
+
+#include <ATen/Error.h>
+#include <atomic>
+#include <stdexcept>
+
+namespace c10 {
+
+/**
+ * intrusive_ptr<T> is an alternative to shared_ptr<T> that has better
+ * performance because it does the refcounting intrusively
+ * (i.e. in a member of the object itself).
+ * Your class T needs to inherit from intrusive_ptr_target to allow it to be
+ * used in an intrusive_ptr<T>.
+ */
+
+class intrusive_ptr_target {
+  // Note [Weak references for intrusive refcounting]
+  // ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+  // Here's the scheme:
+  //
+  //  - refcount == number of strong references to the object
+  //    weakcount == number of weak references to the object,
+  //      plus one more if refcount > 0
+  //    An invariant: refcount > 0  =>  weakcount > 0
+  //
+  //  - THStorage stays live as long as there are any strong
+  //    or weak pointers to it (weakcount > 0, since strong
+  //    references count as a +1 to weakcount)
+  //
+  //  - finalizers are called and data_ptr is deallocated when refcount == 0
+  //
+  //  - Once refcount == 0, it can never again be > 0 (the transition
+  //    from > 0 to == 0 is monotonic)
+  //
+  //  - When you access THStorage via a weak pointer, you must
+  //    atomically increment the use count, if it is greater than 0.
+  //    If it is not, you must report that the storage is dead.
+  //
+  mutable std::atomic<size_t> refcount_;
+  mutable std::atomic<size_t> weakcount_;
+
+  template <typename T, typename NullType>
+  friend class intrusive_ptr;
+  template <typename T, typename NullType>
+  friend class weak_intrusive_ptr;
+
+ protected:
+  // protected destructor. We never want to destruct intrusive_ptr_target*
+  // directly.
+  virtual ~intrusive_ptr_target() {
+    AT_CHECK(
+        refcount_.load() == 0,
+        "Tried to destruct an intrusive_ptr_target that still has intrusive_ptr to it");
+    AT_CHECK(
+        weakcount_.load() == 0,
+        "Tried to destruct an intrusive_ptr_target that still has weak_intrusive_ptr to it");
+  }
+
+  constexpr intrusive_ptr_target() noexcept : refcount_(0), weakcount_(0) {}
+
+ private:
+  /**
+   * This is called when refcount reaches zero.
+   * You can override this to free expensive resources.
+   * There might still be weak references, so your object might not get
+   * destructed yet, but you can assume the object isn't used anymore,
+   * i.e. no more calls to methods or accesses to members (we just can't
+   * destruct it yet because we need the weakcount accessible).
+   *
+   * Even if there are no weak references (i.e. your class is about to be
+   * destructed), this function is guaranteed to be called first.
+   * However, if you use your class for an object on the stack that is
+   * destructed by the scope (i.e. without intrusive_ptr), this function will
+   * not be called.
+   */
+  virtual void free() const {}
```

Comment:
free seems like a worse method name than `release_resources`. The name `free` might suggest that you need to `delete this` in the implementation, but that's not true.